XenAdmin/Dialogs/NewTagDialog.cs

```diff
@@ -422,42 +223,79 @@ public void Toggle()
 
             public void Toggle(CheckState stateToToggleFrom)
             {
-                StateImageIndex = Convert.ToInt32(!Convert.ToBoolean(stateToToggleFrom));
+                Checked = stateToToggleFrom == CheckState.Checked ? CheckState.Unchecked : CheckState.Checked;
             }
 
-            public new CheckState Checked
+            public CheckState Checked
             {
                 get
                 {
-                    if (StateImageIndex == 0)
-                    {
+                    var value = _cellCheckState.Value;
+                    if (value == null)
                         return CheckState.Unchecked;
-                    }
-                    else if (StateImageIndex == 1)
-                    {
-                        return CheckState.Checked;
-                    }
-                    else
-                    {
-                        return CheckState.Indeterminate;
-                    }
-                }
-                set
-                {
-                    if (value == CheckState.Unchecked)
-                    {
-                        StateImageIndex = 0;
-                    }
-                    else if (value == CheckState.Checked)
-                    {
-                        StateImageIndex = 1;
-                    }
-                    else
-                    {
-                        StateImageIndex = 2;
-                    }
+
+                    if (value is CheckState)
+                        return (CheckState)value;
+
+                    if (value is bool)
+                        return (bool)value ? CheckState.Checked : CheckState.Unchecked;
+
+                    return CheckState.Indeterminate;
                 }
+                set { _cellCheckState.Value = value; }
+            }
+
+            public string Text
+            {
+                get { return _cellTag.Value.ToString(); }
+                set { _cellTag.Value = value; }
+            }
+
+            public int CompareTo(TagsDataGridViewRow other)

```

Comment:
I've just realised that the gridview's rows are not clickable, which means sorting is not taking place. This seems to have been a bug in the original code too. The way to fix it is to set SortMode to Automatic on each column and handle the datagridvuiew's SortCompare method (see how this is used elsewhere in XenCenter, e.g. the class BugToolPageSelectCapabilities.cs is a good example).